### PR TITLE
feat(document-index): add hybrid_index to index configuration

### PR DESCRIPTION
**Pull request by TilTheunissenAA**
# Description
No description.

## Before Merging
 - [ ] Review the code changes
    - Unused print / comments / TODOs
    - Missing docstrings for functions that should have them
    - Consistent variable names
    - ...
 - [x] Update `changelog.md` if necessary
 - [x] Commit messages should contain a semantic [label](https://gist.github.com/joshbuchea/6f47e86d2510bce28f8e7f42ae84c716) and the ticket number
   - Consider squashing if this is not the case

### CODE REVIEW

File: tests/connectors/document_index/test_document_index.py [hunk 1]
```diff
@@ -200,6 +201,27 @@ def test_document_index_searches_asymmetrically(
...
+        max_results=10,
+        min_score=0.0,
+    )
+    search_results = document_index.search(
+        document_path.collection_path, "asymmetric-hybrid", search_query
+    )
+
+    assert "Poison in food is bad" in search_results[0].section
+    assert "All people like pasta" in search_results[1].section
+    assert "Chairs are not food" in search_results[2].section
```
Comment (Michael-JB) [inline tests/connectors/document_index/test_document_index.py:222]: Challenge accepted

File: tests/connectors/document_index/test_document_index.py [hunk 2]
```diff
@@ -200,6 +201,27 @@ def test_document_index_searches_asymmetrically(
...
+    document_index: DocumentIndexClient, collection_path: CollectionPath
+) -> None:
+    document_path = DocumentPath(
+        collection_path=collection_path,
+        document_name="test_document_index_hybrid_search_combines_semantic_and_keyword_search",  # is always there
+    )
+    search_query = SearchQuery(
+        query="Which food do humans like?",
+        max_results=10,
+        min_score=0.0,
```
Comment (Michael-JB) [inline tests/connectors/document_index/test_document_index.py:214]: Maybe not one for this PR, but we want to discourage use of `min_score` with searches over hybrid indexes (the re-ranked score is not very useful...), and a min_score of 0.0 does nothing anyway. We should probably make it an optional parameter that defaults to 0.

**Review by Michael-JB — APPROVED** [review]

File: tests/connectors/document_index/test_document_index.py (hunk 2 above)
Comment (Michael-JB) [reply tests/connectors/document_index/test_document_index.py:214]: (I'll raise a task for this)

Comment (Michael-JB) [reply tests/connectors/document_index/test_document_index.py:214]: (PR for this: https://github.com/Aleph-Alpha/intelligence-layer-sdk/pull/1053)

File: src/intelligence_layer/connectors/document_index/document_index.py [hunk 3]
```diff
@@ -36,11 +36,13 @@ class IndexConfiguration(BaseModel):
         chunk_overlap: The maximum number of tokens of overlap between consecutive chunks. Must be
             less than `chunk_size`.
         chunk_size: The maximum size of the chunks in tokens to be used for the index.
+        hybrid_index: If set to "bm25", combine vector search and keyword search (bm25) results.
     """
 
     embedding_type: Literal["symmetric", "asymmetric"]
     chunk_overlap: int = Field(default=0, ge=0)
     chunk_size: int = Field(..., gt=0, le=2046)
+    hybrid_index: Literal["bm25"] | None = None
```
Comment (NiklasKoehneckeAA) [inline src/intelligence_layer/connectors/document_index/document_index.py:45]: are you planning to support more hybrid indices in the future?

File: tests/connectors/document_index/test_document_index.py (hunk 2 above)
Comment (NiklasKoehneckeAA) [reply tests/connectors/document_index/test_document_index.py:214]: Should we then add this into the docstring or give a warning for this? Im not quite sure of the implications and would let you decide

File: src/intelligence_layer/connectors/document_index/document_index.py (hunk 3 above)
Comment (Michael-JB) [reply src/intelligence_layer/connectors/document_index/document_index.py:45]: This is a possibility, yes.

File: tests/connectors/document_index/test_document_index.py (hunk 2 above)
Comment (Michael-JB) [reply tests/connectors/document_index/test_document_index.py:214]: There are two things here:
- min score of 0.0 does nothing: this is the case already and I don't think this needs documenting; given a score is documented as being between 0 and 1, it is intuitive that filtering by >= 0 will always yield a full result set.
- score is not very useful for hybrid indices: this should now be sufficiently documented in the DocumentIndexRetriever once https://github.com/Aleph-Alpha/intelligence-layer-sdk/pull/1053 is merged. We also document this in the DI OpenAPI spec, so I see no need to modify the IL DI client as well.

File: src/intelligence_layer/connectors/document_index/document_index.py (hunk 3 above)
Comment (TilTheunissenAA) [reply src/intelligence_layer/connectors/document_index/document_index.py:45]: We want to be able to add new ones.

**Pull request merged**